_drafts/firmware-code-size-2.md

```diff
@@ -0,0 +1,303 @@
+---
+title: "Shrink your Firmware with GCC"
+description: "A walk through GCC configuration options to optimize firmware code size"
+author: francois
+---
+
+## Introduction
+This is the second post in our Firmware Code Size Optimization series. Last time, we talked about measuring code size as a precondition to improving it. Now we’ll dive leap from planning into action!
+
+<!-- excerpt start -->
+In this post, we will review compiler options that we can use to reduce firmware code size. We will focus on `arm-none-eabi-gcc`, the GCC compiler used for ARM-based microcontroller, though most of the compile-time flags we will cover are available in other GCC flavors as well as in Clang.
+<!-- excerpt end -->
+
+You would think that a single flag could be used for “make this as small as possible”, but unfortunately it isn’t so. Instead, code size optimization involves complicated trade-offs that must be considered on a case by case basis.
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Setting the stage
+In order to get measurable improvements in code size, we need a reasonably large code-base to begin with. To that end, we’ll use example code from [ChibiOS](http://chibios.org/dokuwiki/doku.php), a free and open source RTOS. Specifically, we’ll use their FatFS+USB example for STM32F1 MCUs.
+
+You can reproduce our setup simply by cloning the project:
+```
```

Comment:
Might be worth mentioning what gcc toolchain you are using 

```
# git remote -v
origin	git@github.com:ChibiOS/ChibiOS.git (fetch)
origin	git@github.com:ChibiOS/ChibiOS.git (push)
(.venv) chrisc: ~/dev/ChibiOS/
# git log -1
commit 20742714c0099084e5d63a7c6d47020bde2e81e7 (HEAD -> master, origin/master, origin/HEAD)
Author: Giovanni Di Sirio <gdisirio@gmail.com>
Date:   Sun Aug 11 12:13:52 2019 +0000

    Removed old OSAL implementations.

    git-svn-id: svn://svn.code.sf.net/p/chibios/svn/trunk@12949 27425a3e-05d8-49a3-a47f-9c15f0e5edd8
```

hrm I'm  hitting:

```
# make USE_LINK_GC=no USE_LTO=no
Compiling fatfs_diskio.c
../../../os/various/fatfs_bindings/fatfs_diskio.c:10:10: fatal error: ff.h: No such file or directory
 #include "ff.h"
          ^~~~~~
compilation terminated.
make: *** [build/obj/fatfs_diskio.o] Error 1
```

Reply:
figured it out ... had to do this first:

```
(.venv) chrisc: ~/dev/ChibiOS/ext/
# brew install p7zip
# 7za x fatfs-0.13c_patched.7z
```

Reply:
Added a line to instructions about 7za